presentation/frontend/src/presentation-frontend/StreamedResponseGenerator.ts

```diff
@@ -0,0 +1,144 @@
+   */
+  public get total(): number {
+    return this._total;
+  }
+
+  /**
+   * Fetches and caches the first batch.
+   * This function can be called in order to retrieve the total items count.
+   * @returns response for the first page.
+   */
+  public async fetchFirstBatch(): Promise<PagedResponse<TPagedResponseItem>> {
+    if (this._firstBatch) {
+      return this._firstBatch;
+    }
+
+    const start = this._props.paging?.start ?? 0;
+    const batchSize = this._props.paging?.size ?? 0;
+    this._firstBatch = await this._props.getBatch({ start, size: batchSize }, 0);
+    this._total = this._firstBatch.total;
+    return this._firstBatch;
+  }
+
+  /**
+   * Fetches items and collects to an array.
+   * If page size is specified in the properties, then it will return only the
+   */
+  public async getItems(): Promise<TPagedResponseItem[]> {
+    const result = new Array<TPagedResponseItem>();
+    for await (const value of this.itemsIterator) {
+      result.push(value);
+    }
+    return result;
+  }
+
+  /** Async iterator of all items. */
+  public get itemsIterator(): AsyncIterableIterator<TPagedResponseItem> {
+    return eachValueFrom(this.items);
+  }
+
+  /**
+   * RXJS observable of items.
+   * Items count will be limited to the page size, if one is specified in the configuration.
+   */
+  public get items(): Observable<TPagedResponseItem> {
+    return this._batches.pipe(mergeAll());
+  }
+
+  private get _batches(): Observable<TPagedResponseItem[]> {
+    const pageStart = this._props.paging?.start ?? 0;
+    const parallelism = this._props.parallelism;
+    const pageSize = this._props.paging?.size;
+
+    return from(this.fetchFirstBatch()).pipe(
+      mergeMap((response) => {
+        // If there are no items, return a single empty page.
+        if (this._total === 0) {
+          return of([]);
+        }
+
+        // If the response is empty, something went wrong.
+        const receivedItemsLength = response.items.length;
+        if (!receivedItemsLength) {
+          this.handleEmptyPageResult(pageStart);
+        }
+
+        const totalItemsToFetch = this._total - pageStart;
+        if (receivedItemsLength === totalItemsToFetch) {
+          return of(response.items);
+        }
+
+        let itemsToFetch: number;
+        let batchSize: number;
+        if (pageSize) {
+          itemsToFetch = Math.min(totalItemsToFetch, pageSize) - receivedItemsLength;
+          batchSize = Math.min(pageSize, receivedItemsLength);
+        } else {
+          itemsToFetch = totalItemsToFetch - receivedItemsLength;
+          batchSize = receivedItemsLength;
+        }
+
+        const remainingBatches = Math.ceil(itemsToFetch / batchSize);
+
+        // Return the first page and then stream the remaining ones.
+        return concat(
+          of(response.items),
+          range(1, remainingBatches).pipe(
+            mergeMap(async (idx) => {
```

Comment:
I believe `mergeMap` may return results in order different than the requests are made, e.g. response for request 3 may return earlier than for request 2, and in that case the 3rd batch is going to be emitted before the 2nd. Need a test for this.

Reply:
That's right. Added the logic for aggregating the results in order and an a test